src/managedwriter/writer_client.ts

```diff
@@ -0,0 +1,418 @@
+          ? this._client.appendRows(clientOptions)
+          : this._client.appendRows(),
+      };
+      const managedStream = new ManagedStream(
+        fullStreamId,
+        streamConnection,
+        streamType,
+        protoDescriptor,
+        this
+      );
+      this._connections.connectionList.push(managedStream);
+      this._connections.connections[`${streamId}`] = managedStream;
+      return managedStream;
+    } catch (err) {
+      throw new Error('managed stream connection failed:' + err);
+    }
+  }
+
+  private async resolveStreamId(
+    streamId?: string,
+    streamType?: WriteStreamType,
+    destinationTable?: string
+  ): Promise<string> {
+    if (streamId && streamId !== '') {
+      if (streamId === DefaultStream) {
+        if (destinationTable !== '') {
+          return `${destinationTable}/streams/_default`;
+        } else {
+          throw new Error('destinationTable needed if DefaultStream informed');
+        }
+      }
+      return streamId;
+    }
+    if (streamType && destinationTable) {
+      streamId = await this.createWriteStream({streamType, destinationTable});
+      return streamId;
+    }
+    throw new Error(
+      'streamType and destinationTable required to create write stream'
+    );
+  }
+
+  async batchCommitWriteStream(
+    req: BatchCommitWriteStreamsRequest
+  ): Promise<BatchCommitWriteStreamsResponse> {
+    await this.initialize();
+    const [res] = await this._client.batchCommitWriteStreams(req);
+    return res;
+  }
+
+  async close() {
+    this._connections.connectionList.map(ms => {
+      return ms.close();
+    });
+  }
+
+  async finalize(): Promise<
+    protos.google.cloud.bigquery.storage.v1.IFinalizeWriteStreamResponse['rowCount']
+  > {
+    const rowCounts = await Promise.all(
+      this._connections.connectionList.map(ms => {
+        return ms.finalize();
+      })
+    );
+
+    return rowCounts.reduce((total: number, rowCount) => {
+      const rowCountNum = Number.parseInt(`${rowCount}`, 10);
+      return total + rowCountNum;
+    }, 0);
+  }
+}
+
+class PendingWrite {
+  private result?: AppendRowsResponse;
+  private promise: Promise<AppendRowsResponse>;
+  private resolveFunc?: (response: AppendRowsResponse) => void;
+  private rejectFunc?: (reason?: protos.google.rpc.IStatus) => void;
+
+  constructor() {
+    this.promise = new Promise((resolve, reject) => {
+      this.resolveFunc = resolve;
+      this.rejectFunc = reject;
+    });
+  }
+
+  _setResult(result: AppendRowsResponse) {
+    this.result = result;
+  }
+
+  _markDone() {
+    if (this.result) {
+      if (this.result.error) {
+        this.rejectFunc && this.rejectFunc(this.result.error);
+      }
+      this.resolveFunc && this.resolveFunc(this.result);
+    }
+    this.rejectFunc && this.rejectFunc(new Error('ended with no status'));
+  }
+
+  getResult(): Promise<AppendRowsResponse> {
+    return this.promise;
+  }
+}
+
+export class ManagedStream {
```

Comment:
This is your chance to correct a potentially awkward naming from go. Maybe we just call a managedstream a writer?